### ISSUE TITLE
Producer: throw out error message "Local: Message timed out"

### ISSUE THREAD

**Issue by wbtwins** [edenhill/librdkafka#405]
hi,edenhill
   We downloaded librdkafka-master and deployed it on the centos.it would throw out  "Local: Message timed out" when we run example/example_rdkafka in producer Mode.Following is our command line:
./rdkafka_example -P -t test -p 0  -b 192.168.84.183 -d all

It would throw out message "local message timeout" after the inputed stuff  is time out for delivery.

we traced the code,and found that the message had been enquened,we would print the message in 
rktp->rktp_msgq->rhq_>first->rkm_payload

we also tried rdkafka_performance and other binary in test,it had the same result ,no matter we took jafka-1.4.0 or kafka_2.9.1-0.8.2.2 as our server side.

The config of rdkafka we used is default.
 
Is there any other condition we should set in our environment?

[root@localhost examples]# ./rdkafka_example -P -t test -p 0  -b 192.168.84.183 -d all
1446518837.618 RDKAFKA-7-BROKER: rdkafka#producer-0: 192.168.84.183:9092/1: Added new broker with NodeId 1
1446518837.618 RDKAFKA-7-TOPIC: rdkafka#producer-0: New local topic: test
1446518837.618 RDKAFKA-7-BRKMAIN: rdkafka#producer-0: 192.168.84.183:9092/1: Enter main broker thread
1446518837.618 RDKAFKA-7-CONNECT: rdkafka#producer-0: 192.168.84.183:9092/1: broker in state INIT connecting
% Type stuff and hit enter to send
1446518837.621 RDKAFKA-7-CONNECTED: rdkafka#producer-0: 192.168.84.183:9092/1: connected to <unsupported:inet>
1446518837.621 RDKAFKA-7-STATE: rdkafka#producer-0: 192.168.84.183:9092/1: Broker changed state INIT -> UP
1446518837.621 RDKAFKA-7-METADATA: rdkafka#producer-0: 192.168.84.183:9092/1: Request metadata for all topics: connected
1446518837.621 RDKAFKA-7-METADATA: rdkafka#producer-0: 192.168.84.183:9092/1: Requesting metadata for all topics
1446518837.621 RDKAFKA-7-SEND: rdkafka#producer-0: 192.168.84.183:9092/1: Sent MetadataRequest (25 bytes, CorrId 1)
1446518838.618 RDKAFKA-7-METADATA: rdkafka#producer-0: 192.168.84.183:9092/1: Request metadata for test: leader query
1446518838.618 RDKAFKA-7-METADATA: rdkafka#producer-0: 192.168.84.183:9092/1: Request metadata: scheduled: not in broker thread
1446518838.623 RDKAFKA-7-METADATA: rdkafka#producer-0: 192.168.84.183:9092/1: Request metadata for test: leader query
1446518838.623 RDKAFKA-7-METADATA: rdkafka#producer-0: 192.168.84.183:9092/1: Requesting metadata for known topics
1446518838.623 RDKAFKA-7-SEND: rdkafka#producer-0: 192.168.84.183:9092/1: Sent MetadataRequest (31 bytes, CorrId 2)
ksdfjksdjklfjklsdjfkljsdkljfklsdjkfljsdkjfksdjfkljsdkfj
% Sent 55 bytes to topic test partition 0
sdfkjsdkjfkldsjklfjkdsljfkldsjklfjksdl
% Sent 38 bytes to topic test partition 0
1446518898.690 RDKAFKA-7-METADATA: rdkafka#producer-0: 192.168.84.183:9092/1: ===== Received metadata from 192.168.84.183:9092/1 =====
1446518898.690 RDKAFKA-4-METADATA: rdkafka#producer-0: 192.168.84.183:9092/1: Metadata request failed: Local: Message timed out
1446518898.690 RDKAFKA-7-REQTMOUT: rdkafka#producer-0: 192.168.84.183:9092/1: Timed out 1 requests
1446518899.691 RDKAFKA-7-METADATA: rdkafka#producer-0: 192.168.84.183:9092/1: ===== Received metadata from 192.168.84.183:9092/1 =====
1446518899.691 RDKAFKA-4-METADATA: rdkafka#producer-0: 192.168.84.183:9092/1: Metadata request failed: Local: Message timed out
1446518899.691 RDKAFKA-7-REQTMOUT: rdkafka#producer-0: 192.168.84.183:9092/1: Timed out 1 requests
1446518900.633 RDKAFKA-7-METADATA: rdkafka#producer-0: 192.168.84.183:9092/1: Request metadata for test: leader query
1446518900.633 RDKAFKA-7-METADATA: rdkafka#producer-0: 192.168.84.183:9092/1: Request metadata: scheduled: not in broker thread
1446518900.691 RDKAFKA-7-METADATA: rdkafka#producer-0: 192.168.84.183:9092/1: Request metadata for test: leader query
1446518900.691 RDKAFKA-7-METADATA: rdkafka#producer-0: 192.168.84.183:9092/1: Requesting metadata for known topics
1446518900.692 RDKAFKA-7-SEND: rdkafka#producer-0: 192.168.84.183:9092/1: Sent MetadataRequest (31 bytes, CorrId 3)

**Comment by wbtwins** [comment]
edenhill,
   would the rdkafka client  support Jafka server?or it can only communicate with Kafka server?
   

Thanks in advance！

**Comment by wbtwins** [comment]
BTW:
we can find that all the produced message are in the quenue,but they are not deliveried on time.

./rdkafka_performance -P -t test2 -s 100 -c 500000 -m "this is test" -S 1 -a 2  -b 192.168.84.183
% 1 messages produced (100 bytes), 0 delivered (offset 0, 0 failed) in 0ms: 0 msgs/s and 0.00 Mb/s, 0 produce failures, 1 in queue, no compression
% 500000 messages produced (50000000 bytes), 0 delivered (offset 0, 0 failed) in 1000ms: 0 msgs/s and 0.00 Mb/s, 0 produce failures, 500000 in queue, no compression
% 500000 messages produced (50000000 bytes), 0 delivered (offset 0, 0 failed) in 2001ms: 0 msgs/s and 0.00 Mb/s, 0 produce failures, 500000 in queue, no compression
% 500000 messages produced (50000000 bytes), 0 delivered (offset 0, 0 failed) in 3002ms: 0 msgs/s and 0.00 Mb/s, 0 produce failures, 500000 in queue, no compression
% 500000 messages produced (50000000 bytes), 0 delivered (offset 0, 0 failed) in 4003ms: 0 msgs/s and 0.00 Mb/s, 0 produce failures, 500000 in queue, no compression
% 500000 messages produced (50000000 bytes), 0 delivered (offset 0, 0 failed) in 5004ms: 0 msgs/s and 0.00 Mb/s, 0 produce failures, 500000 in queue, no compression
% 500000 messages produced (50000000 bytes), 0 delivered (offset 0, 0 failed) in 6005ms: 0 msgs/s and 0.00 Mb/s, 0 produce failures, 500000 in queue, no compression
% 500000 messages produced (50000000 bytes), 0 delivered (offset 0, 0 failed) in 7005ms: 0 msgs/s and 0.00 Mb/s, 0 produce failures, 500000 in queue, no compression
% 500000 messages produced (50000000 bytes), 0 delivered (offset 0, 0 failed) in 8006ms: 0 msgs/s and 0.00 Mb/s, 0 produce failures, 500000 in queue, no compression
% 500000 messages produced (50000000 bytes), 0 delivered (offset 0, 0 failed) in 9007ms: 0 msgs/s and 0.00 Mb/s, 0 produce failures, 500000 in queue, no compression
% 500000 messages produced (50000000 bytes), 0 delivered (offset 0, 0 failed) in 10009ms: 0 msgs/s and 0.00 Mb/s, 0 produce failures, 500000 in queue, no compression
% 500000 messages produced (50000000 bytes), 0 delivered (offset 0, 0 failed) in 11009ms: 0 msgs/s and 0.00 Mb/s, 0 produce failures, 500000 in queue, no compression
% 500000 messages produced (50000000 bytes), 0 delivered (offset 0, 0 failed) in 12009ms: 0 msgs/s and 0.00 Mb/s, 0 produce failures, 500000 in queue, no compression
% 500000 messages produced (50000000 bytes), 0 delivered (offset 0, 0 failed) in 13009ms: 0 msgs/s and 0.00 Mb/s, 0 produce failures, 500000 in queue, no compression
^C% 500000 messages produced (50000000 bytes), 0 delivered (offset 0, 0 failed) in 14010ms: 0 msgs/s and 0.00 Mb/s, 0 produce failures, 500000 in queue, no compression

no message is delivered.

**Comment by edenhill** [comment]
The Metadata requests are timing out for some reason:
`1446518898.690 RDKAFKA-7-REQTMOUT: rdkafka#producer-0: 192.168.84.183:9092/1: Timed out 1 requests`

Can you check the broker logs on 192.168.84.183 to see why?

**Comment by wbtwins** [comment]
we checked the log ,the server side get package which only contains topic.
To make sure it's not the problem with platform.we also tried on ubuntu.unfortunately,it also failed.

we use gdb to trace the rdkafka_example,found that briker thread only never put the messge in rkt_topic into buf in the function rd_kafka_broker_metadata_req_op:

	if (rko->rko_rkt || !rko->rko_all_topics) {
		/* Just our locally known topics */

		TAILQ_FOREACH(rkt, &rkb->rkb_rk->rk_topics, rkt_link) {
			int tlen;
			if (rko->rko_rkt && rko->rko_rkt != rkt)
				continue;
			tlen = RD_KAFKAP_STR_SIZE(rkt->rkt_topic);
			memcpy(buf+of, rkt->rkt_topic, tlen);
			of += tlen;
		}
		rd_kafka_unlock(rkb->rkb_rk);
	}

the thread can goes into above condition.  So when thread run into function rd_kafka_send,it would fail to check  rkbuf->rkbuf_of which is zero.

Thanks in advance!

**Comment by wbtwins** [comment]
sorry for above message,  "the thread can goes into above condition." should be ""the thread can not goes  into above condition.""

**Comment by edenhill** [comment]
I'm not sure I follow, are you saying the metadata request is malformed?
Can you provide the relevant broker logs for this?

**Comment by wbtwins** [comment]
I mean that  the message in the payload is not delivered ,just only the topics

**Comment by edenhill** [comment]
Could you provide the broker logs?

**Comment by wbtwins** [comment]
OK, I would provide the log asap.
BTW:
   1.Do we only need to link librdkafka to our own program to use the client?
   2.should  there  be any special configuration  to make the client run correct?
   3. Is there any special configuration to be set on kafka server side?

**Comment by edenhill** [comment]
You link your application either dynamically or statically with librdkafka.

There is no special configuration required to run the client, all it needs is the initial set of bootstrap broker(s): `metadata.broker.list=mybroker`

No special config is needed on the broker.

**Comment by wbtwins** [comment]
Thanks . I just retried on my computer at home .It works well. 
I deployed kafka server in windows,and run two rdkafka_example instances in the 64-bit ubuntu.
One is producer,and another is  consumer.It work well  .Consumer can get the message from producer.


Producer:
wb@ubuntu:/mnt/hgfs/librdkafka-master/examples$ ./rdkafka_example -P -t WTC -p 0 -b 192.168.1.7 
% Type stuff and hit enter to send
1446556560.057 RDKAFKA-5-TOPICBRK: rdkafka#producer-0: Topic WTC [0] migrated to unknown broker 0: requesting metadata update
test for kafka                                                                                           
% Sent 14 bytes to topic WTC partition 0


Consumer:
% Message (offset 58, 14 bytes):
Message Payload hexdump (14 bytes):
00000000: 74 65 73 74 20 66 6f 72 20 6b 61 66 6b 61        test for kafka  
% Consumer reached end of WTC [0] message queue at offset 59



But it would not work well in computer in company which works on centOs.
I don't think the platform would cause the failure. Need to double check the environment.

THanks for your help
